Floating layout lays out windows with USPosition/PPosition

tcld opened the pull request:
Most floating windows (that were not manually put into floating mode)
have either USPosition or PPosition already set (sometimes both).
Therefore we cannot reliably check these flags to see whether the
floating layout should lay out a window.

Most naturally floating windows already come with a valid position,
which will be respected. The floating layout will lay out the windows
anyway. Windows with invalid coordinates are repositioned first.

Fixes #1893 and #1913.

Every software I have tested worked great, most importantly the above issue have been fixed. It is possible, though, that some other pieces of software might still show weird behaviour.

tcld commented on libqtile/layout/floating.py:
```diff
@@ -181,31 +181,35 @@ def compute_client_position(self, client, screen_rect):
+            y = center_y - client.height / 2
+
+            # don't go off the right...
+            x = min(x, screen_rect.x + screen_rect.width - client.width)
```
I made a small change here, since it seems more sensible to have the entire window on screen than just requiring it to be adjacent to it.

tych0 replied:
Can you add this reasoning to the commit message itself?

tcld replied:
OK, I reworded the commit message.

tych0 commented:
Huh? We don't want to just delete this code... perhaps we should add logic to figure out the right screen. But ignoring USPosition is not correct either.

tcld commented:
The function in the `Window`-class isn't used by anything else. Or are you talking about the lines I removed from the floating layout?

Also, this is not just about the screen. The other issue about child-windows and popups seems to be solved by this, too (I haven't performed a proper test comparing the master branch and just this PR yet).

tych0 commented:
> Or are you talking about the lines I removed from the floating layout?

Yes :). We should not ignore USPosition.

tcld commented:
Well, this PR doesn't entirely ignore the flags - there are two differences from before:
1. Instead of checking whether the flags are set, _the coordinates themselves_ are checked. If they don't make sense, they are overwritten. Otherwise we trust the software that set them. This fixes #1913. 
2. For every window with USPosition and/or PPosition set, `_Window.place` is called instead of `_Window.user_placed_window_setup`. This fixes #1893. 

There is probably an approach where we could still check the flags - but the only thing I could see be done differently, is how the windows _without_ either of the flags set would be handled (i.e. always instead of just when the given coordinates are weird). So instead of [the checks here](https://github.com/qtile/qtile/pull/1943/files#diff-475097a9172bb7cb3cc2bc9fe378a383acf1231b6a6bc1ec048e764af439eddcR186) we could do something like this:
```python
if not client.has_user_set_position:
```

What do you think?

tych0 commented:
> Instead of checking whether the flags are set, the coordinates themselves are checked. If they don't make sense, they are overwritten.

Right, I think this is incorrect. If a window specified its position, we should do our best to put it there.

What I'm suggesting is that the coordinates are perhaps relative to the screen the window is on, and the bug is that we're assuming they're global. But I don't think overwriting the coordinates is correct.

tcld commented:
That's a good idea, it seems that some software provides "global" coordinates, while other software provides local coordinates - that would explain some of the problems I noticed. So overwriting the nonsensical coordinates would be correct after all, it's just the way it is done that should be changed. I'll add another commit tomorrow.

tcld commented:
I reintroduced the `has_user_set_position`-method. If it is `True` *and* the client is on screen, we don't mess with its coordinates. In case it is off screen, the coordinates need to be changed, however.

Either way, the layout will actuall call `place()` on the client, but maybe without changing the coordinates first.

@tych0 : Is this more what you imagined?

tych0 commented on libqtile/layout/floating.py:
```diff
@@ -231,12 +249,6 @@ def configure(self, client, screen_rect):
-        # similar to above but the X11 version, the client may have already
-        # placed itself. let's respect that
-        if client.has_user_set_position():
-            client.user_placed_window_setup(bc, bw)
```
Is there a way we can avoid a place() call if has_user_set_position() is true and we don't have to change the coordinates? It does avoid some flicker.

tcld replied:
Not calling `_Window.place()` seems to be what is causing #1893. I guess that means that the call to `_Window.send_configure_notify`is the important part? Other than that `place` does what `user_placed_window_setup` does, as far as I can see. I'll try to take a closer look.

@m-col : Your PR #1887 touches a lot of this code, too. I'll try to see whether that, too, solves #1893.

tcld replied:
It seems that #1887 also solves #1893. It doesn't adress the cases where floating windows spawn on the wrong screen, however. So I think #1887 would be preferrable over this PR, I could then later try to fix the problem with spawning on the wrong screen separately.

tcld replied:
I put some time into investigating why calling `place` for windows having `USPosition` and/or `PPosition` solves having invisible popups/children (#1893): The critical call needed is that to `xcbq.Window.configure(x, y, width, height)`.

This seems very confusing to me, since the window should already be positioned and visible without that call, right? For the cases described in #1893, however, only _every other_ instance appears visible. There might be an underlying bug in how we grab and manage windows, but I don't think I can find it on my own.

Here is the change that would fix #1893 - maybe somebody else knows why this works and what underlying issue it might be pointing to? (@m-col , @ramnes ?)
```python
# inside layout.floating.Floating.configure
if client.has_user_set_position():
    client.user_placed_window_setup(bc, bw)
    client.window.configure(x=client.x, y=client.y, width=client.width, height=client.height)
```

Any help would be very much appreciated!

tcld commented:
Rebased to current master. Also moved the `on_screen`-method so it can be called externally (could be useful, since the recent changes to the screen change-hook can leave floating windows off screen in some edge cases).

The intended fix with this PR is #1913, #1893 is just an accident.

tych0 commented on libqtile/layout/floating.py:
```diff
@@ -231,12 +249,6 @@ def configure(self, client, screen_rect):
-            client.paint_borders(bc, bw)
-            client.cmd_bring_to_front()
-
         # ok, it's not java and the window itself didn't position it, but users
```
This comment is now wrong, IIUC

tcld replied:
You are correct, thanks for spotting that. I changed the comment, too.

tych0 commented:
Great, thanks for working on this!

tcld commented:
I mixed up the issue numbers: This definitely fixes #1913 and works around #1893, not the other way around. But since both issues should be gone, we can get away with closing both for now. I went ahead and closed #1913.
